Beta release: DX7 synth for the Deluge

bfredl opened the pull request:
based on the [dexed](https://github.com/asb2m10/dexed) and [music-synthesizer-for-android](https://github.com/google/music-synthesizer-for-android)  engine implementation with some adjustments to fit the deluge architecture. (less memory consumption, support variable audio buffer size)

DX7 compatible .syx files (there are many to find on the internet) should be placed on the sdcard in a "DX7/" folder.

https://www.youtube.com/watch?v=b-XPERxaB50

How to use:
- create in init synth sound. subtractive amp env will apply, so increase env1 release to around 30 to hear the release phase of the DX envelopes

- change "osc 1 type" to DX7 and press "select" to enter DX editor

- press LOAD twice inside the "dx editor" to load sounds from syx files. Scroll and select a file using the "select" nob. Then patches can be switched by scrolling the "select" nob (no press needed)

- most parameters can be edited. Try hold "shift" and press around the main (8x16) pads to select parameters.

- the mute/launch buttons can be used to switch on/off operators. green = carrier on, blue = modulator on, red = switched off (Like on the dx7 operator switches are not stored, set level to zero to permanently disable an operator in a saved synth.)

limitations/bugs:

- Editor uses the OLED screen a lot. 7seg might compile and work but showing parameters on the 7SEG display is yet missing.

- cpu usage has not been measured, I have no idea how it compares with the builtin FM engine in terms of polyphony.

- short attack sounds can get "clicky", need to add a builtin fade-in. as a workaround, decrease "rate 1" of all operators to be below 80-85 which still is pretty fast but no click.

- keyboardscreen (if active) is functional in editor but not visible Some functionality is missing tho (like you cannot scroll the keyboard). Audition of Dx7 drums in a kit does not work yet.

- saving/loading sounds works. Such synths does not cause trouble on older firmware, but obviously they will not sound properly.

- TODO: operator and rate scaling uses raw midi note - might want a mode where it tracks transposed pitch as well.

- TODO: per-op integration into mod matrix, to fully support MPE and fun stuff. currently only global pitch + level can be modulated externally

- TODO: send/recieve parameter changes and syx dumps over MIDI with compatible hardware.

bfredl committed:
Beta release: DX7 synth for the Deluge

based on the Dexed/MSFA engine implementation with some adjustments
to fit the deluge architecture. (less memory consumption,
support variable audio buffer size)

DX7 compatible .syx files (there are many to find on the
internet) should be placed on the sdcard in a "DX7/" folder.

Polyphony: using the "modern" engine with NEON kernel (now the default)
gives polyphony up to 56 notes for a raw dx7 voice with all 6 operators
(no FX, no filter, etc.) Expected to be slightly lower with feedback active.

The "vintage" engine with lookup tables to mimic the original DX7
more closely has lower polyphony, roughly the half (~28 notes).
I might look into using NEON with the tables if that is possible.

How to use:
- create in init synth sound. subtractive amp env will apply, so
  increase env1 release to around 30 to hear the release phase of
  the DX envelopes

- change "osc 1 type" to DX7 and press "select" to enter DX editor

- press LOAD twice inside the "dx editor" to load sounds from syx files.
  Scroll and select a file using the "select" nob.
  Then patches can be switched by scrolling the "select" nob (no press needed)

- most parameters can be edited. Try hold "shift" and press around
  the main (8x16) pads to select parameters.

- the mute/launch buttons can be used to switch on/off operators.
  green = carrier on, blue = modulator on, red = switched off
  (Like on the dx7 operator switches are not stored, set level to zero
  to permanently disable an operator in a saved synth.)

limitations/bugs:

- cpu usage has not been measured, I have no idea how it compares with
  the builtin FM engine in terms of polyphony.

- short attack sounds can get "clicky", need to add a builtin fade-in.
  as a workaround, decrease "rate 1" of all operators to be below 80-85
  which still is pretty fast but no click.

- keyboardscreen (if active) is functional in editor but not visible
  Some functionality is missing tho (like you cannot scroll the
  keyboard). Audition of Dx7 drums in a kit does not work yet.

- saving/loading sounds works. Such synths does not cause trouble
  on older firmware, but obviously they will not sound properly.

- TODO: operator and rate scaling uses raw midi note - might want a mode
  where it tracks transposed pitch as well.

- TODO: per-op integration into mod matrix, to fully support MPE and fun stuff.
  currently only global pitch + level can be modulated externally

- TODO: send/recieve parameter changes and syx dumps over MIDI with
  compatible hardware.

litui committed:
Big squash of DBT changes.

See PR #88 for full details.

sapphire-arches committed:
chore: remove evil gotos

the compiler can generate good code from the readable version

stellar-aria committed:
Building for 7seg

eduardoj commented:
I guess you wanted to create a pull request against upstream... (SynthstromAudible/DelugeFirmware)

bfredl commented:
Eventually yes, but this is not ready for that yet.

alter-alter committed:
Quantize

litui committed:
Temp removal of PyCortexMDebug

alter-alter committed:
Master Compressor

Master Compressor

Formatted

alter-alter committed:
Merge branch 'master-compressor' into master-comp

alter-alter committed:
Master Compressor

Applied clang format

litui committed:
Whole lot more stuff.

Will update on the PR.

litui committed:
Incremented DBT-Toolchain required version.

litui committed:
Reverting to the community .cproject config so I don't break e2 just yet.

litui committed:
Fixed autotargeting when no targets entered.

litui committed:
Bug fix to pseudolabel detection.

litui committed:
Added detection of detached head state.

litui committed:
Cleaned up git wheel installation routine.

- Only searches for "certifi" as the first wheel to install without
  hardcoding it.
- Uses requirements.txt file and -f to point at the wheel directory.
  This should mean being able to add other requirements to this file
  rather than modifying the dbt scripts in future.

litui committed:
Fixed e2_orig_prefix oversight.

litui committed:
Version string variable substitution.

Apparently I forgot to check my changes after the directory
restructure went through. Added my preprocessor variable
substitution back in and removed the needless DBT.h header file.

stellar-aria committed:
Merge remote-tracking branch 'upstream' into feature/menu-item-rework

stellar-aria committed:
Both OLED and 7seg compiling

litui committed:
Added DBT instructions to top of README.md.

stellar-aria committed:
Menus as a separate cpp

stellar-aria committed:
fix infinite recursion

topisani committed:
Implement Horizontal shift for audio clips

stellar-aria committed:
Remove oneline ifs (readability)

stellar-aria committed:
add base clang-tidy and clangd configurations

topisani committed:
refactor: add getCurrentClip() to clip_view.cpp

topisani committed:
Unify Audio- and instrument clip horizontal shift

topisani committed:
Update cluster reasons on audio clip horizontal shift

stellar-aria committed:
Remove oneline-ifs for OLED

stellar-aria committed:
Run clang-format over source tree

stellar-aria committed:
Add format task

topisani committed:
Disallow shiftHorizontally in more cases, stop audio while changing values

I don't actually know if the latter is needed, but the sample_marker_editor
does it when changing these values, so it seems reasonable to do as well.

I also removed the "CANT" message when shiftHorizontally fails, as i felt it
to be a bit redundant - usually it fails for trivial reasons, such as start
or end reached

stellar-aria committed:
Merge branch 'community' of https://github.com/SynthstromAudible/DelugeFirmware into feature/no-oneline-ifs

stellar-aria committed:
Format new source from upstream

stellar-aria committed:
Style compliance

stellar-aria committed:
Remove contrib/format as per @bobtwinkles' suggestion

m-m-adams committed:
global CC support

litui committed:
Formatted?

New to this editorconfig/clang-format thing.

litui committed:
Merge branch 'community' into feature/dbt-e2-toolchain-changes

stellar-aria committed:
Merge remote-tracking branch 'upstream' into feature/menu-item-rework

stellar-aria committed:
Fix bugs found by @bobtwinkles

stellar-aria committed:
Fix bug found by @bfredl

topisani committed:
Add audio clip horizontal shift to communityfeatures.md

stellar-aria committed:
Merge branch 'community' into feature/no-oneline-ifs

stellar-aria committed:
Final little bits

stellar-aria committed:
Remove temp.h (refactoring artifact)

alter-alter committed:
Changed the UI to use the tempo knob

dctucker committed:
refactor uart to button/pad logic

alter-alter committed:
Moved SimpleComp to MasterCompressor, added shortcuts.

stellar-aria committed:
Add licensing information to all newly split files

jamiefaye committed:
Merge branch 'fix-evil-goto' of https://github.com/bobtwinkles/DelugeFirmware into community

jamiefaye committed:
Merge pull request #129 from alter-alter/quantize

Quantize

jamiefaye committed:
Merge branch 'community' into master-comp

jamiefaye committed:
Merge branch 'community' into feature/no-oneline-ifs

jamiefaye committed:
Merge pull request #145 from stellar-aria/feature/menu-item-rework

MenuItem rework (splitting)

alter-alter committed:
Fixed runtime_feature_setting.h

stellar-aria committed:
Merge branch 'community' of https://github.com/SynthstromAudible/DelugeFirmware into feature/no-oneline-ifs

alter-alter committed:
Merge branch 'SynthstromAudible:community' into master-comp

stellar-aria committed:
Add braces to things that changed due to  merge

jamiefaye committed:
Merge pull request #143 from stellar-aria/feature/no-oneline-ifs

Remove oneline ifs (readability)

alter-alter committed:
Merge branch 'SynthstromAudible:community' into master-comp

litui committed:
Merged community changes into dbt-e2-toolchain-changes.

- Modified src/gui/ui/menus.cpp with version substitution code.

litui committed:
DBT suggested fixes.

Removed and ignored openocd.log.

Fixed python formatting.

sapphire-arches committed:
actions: use litui/dbt-toolchain

jamiefaye committed:
Merge pull request #141 from topisani/feature/audio-clip-rotate

Horizontal Shift for Audio Clips

litui committed:
Merge branch 'community' into feature/dbt-e2-toolchain-changes

litui committed:
Merge pull request #156 from bobtwinkles/patch-1

actions: use litui/dbt-toolchain

litui committed:
Repointed docker pull.

- Needed to be pointing to litui/dbt-toolchain

dctucker committed:
Merge branch 'community' into uart-refactor

dctucker committed:
Button and Pad are classes not structs

sapphire-arches committed:
toolchain: Update REQUIRED_VERSION

Stray commit bumped this, revert until the rest of v9 is ready.

sapphire-arches committed:
chore: fix style change after merge ordering issues

alter-alter committed:
Fixed build error and applied clang-format (MasterCompressor)

alter-alter committed:
Merge branch 'SynthstromAudible:community' into master-comp

alter-alter committed:
Applied clang-format

alter-alter committed:
Update REQUIRED_VERSION

dctucker committed:
buttonAction receives Button not int,int

dctucker committed:
add hid::button constants

dctucker committed:
const button now constexpr enum

jamiefaye committed:
Merge pull request #137 from alter-alter/master-comp

Master Compressor

jamiefaye committed:
Merge pull request #160 from bobtwinkles/patch-2

toolchain: Downgrade REQUIRED_VERSION

litui committed:
Merge remote-tracking branch 'origin/community' into feature/dbt-e2-toolchain-changes

litui committed:
Very intentional increase of dbt version to 9.

dctucker committed:
use CAPS names for hid::button enum

dctucker committed:
using namespace hid::button in buttonActions

dctucker committed:
Merge branch 'community' into uart-refactor

litui committed:
Relocated openocd configurations.

The new paths and naming conventions will make it much easier for DBT to handle additions.

Changed the paths in launch files and one-off scripts as well.

dctucker committed:
search/replace .*Button[XY] after community merge

litui committed:
Fixed bugs and refactored targeting.

- Created target_handler for managing all targets, real or aliased.
    - Far more consistent calls and easier to both work with and troubleshoot.
- Added new exit codes

litui committed:
Small refactoring of openocd and gdb targeting.

While I was in the neighbourhood.

m-m-adams committed:
resolve merge conflicts from rename and single line if

jamiefaye committed:
Merge pull request #147 from m-m-adams/global_command_cc

global CC support

jamiefaye committed:
Merge pull request #88 from SynthstromAudible/feature/dbt-e2-toolchain-changes

dbt integration rollup + version string (phase 1)

jamiefaye committed:
Merge pull request #153 from dctucker/uart-refactor

Refactor UART message to button/pad logic

sapphire-arches committed:
workflows: add merge queue triggers

jamiefaye committed:
Merge pull request #164 from bobtwinkles/ci/add-merge-queue-triggers

workflows: add merge queue triggers

stellar-aria committed:
Fix .bss position and put .text inside .reset (#159)

* Fix .bss position and put .text inside .reset

* Add alignment necessary for v0 bootloaders

trappar committed:
Add names to CI workflows (#165)

litui committed:
Add Python Black Formatter as a recommend (#166)

- The out-of-the-box behaviour is being used for DBT

sapphire-arches committed:
remove straggling index-based button access (#167)

sapphire-arches committed:
ci: do not use a matrix for builds (#168)

this should save us about 3 minutes (50%) of total CI runtime by only
downloading the toolchain once and building everything in parallel.

stellar-aria committed:
Convert contents of deluge/ to #pragma once (#171)

bfredl committed:
midi: allow bigger sysex messages to be sent, such as the OLED image (#174)

* midi: allow bigger sysex messages to be sent, such as the OLED image.

Some simplifications were made to the driver code by Rohan under the
assumption we were only going to queue up to 32 usb-midi messages
for a device per "tick" of the main routine.
These had to be reconsidered, like there is now again a callback from the usb driver
interrupt back to midi_engine to check for more data to send.

It doesn't seem to be possible to increase the amount of data sent in
each call to the driver more than 32 messages (or 128 bytes).
Instead, implement a ring buffer for midi data waiting to be sent over USB.
When we get an interrupt for usbSendComplete, we can check the ring
buffer again and restart the low-level send until the ring buffer is
empty.

Implement a sysex request for sending the contents of the OLED display
as a sysex reply message. This needs more work before being
generally useful as an external display, but serves as a test case
for larger sysex messages.

* midi: use MIDIDevice as an "address" for sysex routing instead of indicies

These objects are intended to be used this way. If disconnected these
get disabled, but not pointers into invalid memory

above class MIDIDevice:
> These never get destructed. So we're safe having various Instruments etc holding pointers to them.

* midi: fix for sending as a host to a device

* cleanup and address review comments by m-m-adams

* midi: add third midi port dedicated for sysex

I kept some refactors in the midi device menu even though I didn't add
the new port to the menu right now, as there is no relevant settings.
we could add it later by just adjusting `lowestDeviceNum`

dctucker committed:
add .clang-format-ignore support to task-format (#175)

stellar-aria committed:
Context Menu cleanup (#177)

* Add namespaces to context_menu

* Rename context_menu files

* Update includes

* Move sample_browsers

* Fix includes/compile

* General cleanup for Context Menus

* Introduce new `Sized` class

* Fix currentOption storage for AudioInputSelector

bfredl committed:
runtime features: make style of patch cable strenght selectable (#184)

Addresses #158

drawBar() method is moved unmodified from IntegerContinuous to
Number as it only depends of attributes of the Number base class

stellar-aria committed:
Fix format task (#181)

dctucker committed:
Refactor indicator LEDs (#180)

* byte enum for IndicatorLEDs

* namespace indicator_leds, not IndicatorLEDs

* apply linter

* IndicatorLED is now an enum class

bfredl committed:
refactor: remove unmaintained code for the 40-pad deluge (#185)

While this code certainly has some nostalgic value, the code code
currently does not compile, and is not directly useful to
the community firmware project. The effort to bring this code back to
a workable state would be large, and there is likely a lot of subtle
bugs creeping in even if one were to get it into a compilable taste.

Not to mention, the number of people this effort would benefit and also would
be able to help with the forward-going maintenance is rather small.

For the cases where consulting these differences is useful, e.g. for
reverse engineering, one can look into an older state of the git
repo. For most cases, like when I was trying to get an overview of the
PIC UART protocol, having half of the bufferPICUart() calls being
dead code is just a distraction.

Co-authored-by: Jamie Fenton <jamie@fentonia.com>

sapphire-arches committed:
Make community features menu a regular submenu (#182)

* runtime-features: non-static init code

* menu: refactor runtime feature menus

This makes the runtime feature menu in to a "real" submenu, in
preparation for community settings that need to have their own menu
subclass (e.g. they need to run some custom code immediately when the
setting is changed)

---------

Co-authored-by: Jamie Fenton <jamie@fentonia.com>

PaulFreund committed:
Editor setting improvements (#172)

* Changed .gitattributes to only allow LF endings

* Fixed line endings of CODEOWNERS file

* Ensure all files end with a final newline (as is required by POSIX)

* Started refactoring launch.json

* Load and Run works on Windows now with a little cd hack

jamiefaye committed:
Remove bad line checked in by mistake (#186)

jamiefaye committed:
Revert "Remove bad line checked in by mistake (#186)" (#187)

This reverts commit d6bb4b77b7b2bd28115dc57657411f57c176905b.

jamiefaye committed:
Revert "refactor: remove unmaintained code for the 40-pad deluge (#185)"

This reverts commit 9acfaf469b3c32e3842f491f8b0012184618a91f.

bfredl committed:
fixes for merge conflicts  (#188)

* Revert "Revert "refactor: remove unmaintained code for the 40-pad deluge (#185)""

This reverts commit fb5016a6086cf5ac6c5bb85360c881f8c503f061.

* fix for merge commits

- #180 and #185 had overlapping changes which were incorrectly resolved in favor of 185
- #182 and #184 had a conflict which were (mostly correctly) resolved in favour 182, but the new menu entry then got missing. Re implemented it in the new structure.

bfredl committed:
remove aliases for bufferPICUart and restructure some LEDs stuff (#189)

These aliases were likely meant as a declaration of intention but
in many cases they are just used in a way that doesn't match the alias
name.

Intention is better being communicated by using wrapper functions in the
PadLEDs and IndicatorLEDs namespaces, to separate the high level UI
code from effectively hardware driver code.

Clean-up scroll code. Previously PadLEDs namespace would expose
horizontal scroll, while vertical scroll logic were hidden away in
LoadSongUI. Instead use "horizontal" and "vertical" sub-namespaces
to fit both scroll dimensions into PadLEDs.

suivaht committed:
Fine tempo knob (#178)

* Update runtime_feature_settings.h

add FineTempoKnob to runtime_feature settings

* Update playback_handler.cpp

added changes to enable FineTempoKnob coarse or fine default

* Update runtime_feature_settings.h

FineTempoKnob now defaults to OFF (can be enabled in runtime features menu

* Update playback_handler.cpp

I have cleaned up the code a bit (used 'switch' instead of 'if...else' for the main logic but had trouble trying to replace the 'runtimeFeatureSettings' line with a variable)

* Update playback_handler.cpp

* Update playback_handler.cpp

removed unnecessary comments

* chore: fix formating

---------

Co-authored-by: bobtwinkles <srkoser+github@gmail.com>

kleinstar1 committed:
Change row color in session View #enhancement (#163)

* Update session_view.cpp

Change row color by pressing row & shift & vertical encoder. This is the same shortcut as in clip view.

* Update CommunityFeatures.md

* Update CommunityFeatures.md

* Update session_view.cpp for spell checker

stellar-aria committed:
Fix building for e2 (#191)

bfredl committed:
sysex: add 7SEG over sysex and improve OLED (#192)

Run Lenght Encoding avoids wasting a lot of bandwidth on large empty
regions.

stellar-aria committed:
C++ Allocator-compliant allocator and string (#194)

* Add deluge::memory::fallback_allocator

* Add deluge::string

* Formatting

seangoodvibes committed:
Midi Takeover Mode (#170)

* Midi Pickup Mode

Midi Pickup Mode enables the Deluge to ignore Midi CC Values from Controllers that are out of Sync with the Deluge.

For example, if an External Midi Controller is sending a Midi CC Value of 0 to the Deluge for a Parameter on the Deluge that is currently set to an equivalent Midi CC Value of 127, then the devices are out of sync.

The Deluge will ignore Midi CC Values from the Midi Controller until the Midi Controller has been "reset" / matches the current Parameter Value on the Deluge.

It does this by comparing the incoming Midi CC Value with the Deluge's internal Knob Position (which is derived from the Current Parameter Value).

The Deluge Internal Knob Position can be converted to a Midi CC Value and Vice Versa by subtracting 64 from the Midi CC Value or adding 64 to the Deluge's Knob Position.

If the difference between the Deluge's Internal Knob Position and the Midi CC Value is greater than +/- 1, then the Midi CC Value is ignored the Parameter's Value will not be changed.

The feature can be disabled using the Community Features menu in Settings. on the Deluge.

* Update to Midi Takeover

Updated the code to:
- Move Midi Takeover Mode settings from Community Menu to Midi Menu
- Add Value Scaling mode
- Save Midi Takeover Settings to Midi XML file
- Load Midi Takeover Settings from Midi XML file

* Resolved runtime_feature_settings.h file conflict

Replaced file with latest version from community

* Corrections made based on feedback

Per m-m-adams feedback, I have made the following changes:

1) moved takeover mode description to community features.md

2) replaced * 1000000 multiplication with << 20

3) replaced / 1000000 division with >> 20

* Fixed formatting issues

Corrected formatting issues identified by clang-format style check

* Trying again to fix formatting!

Hopefully this works!

* Fixed clang format issues

Ok now I think I've got it! Let's try this :)! I ran the clang format script this time.

---------

Co-authored-by: Jamie Fenton <jamie@fentonia.com>

alter-alter committed:
Manual slicing (#198)

* Manual slicing

Update slicer.cpp

* Changed to display slice numbers

* Changed to display Transpose value when pressing the ▼▲ encoder

* Transpose popup is now turned off when a new pad is pressed

* Applied clang-format

PaulFreund committed:
Emit linker error if someone tries to use allocating new (#199)

* Give linker error on use of non address new and fix errors that arise because of this

* Forgot formatting

* Fixed another wrong formatting

* Restored original algorithm state, fixed new signature

alter-alter committed:
Fix MasterCompressor by considering masterVolumeAdjustment value (#200)

* Fix MasterCompressor by considering masterVolumeAdjustment value

* Fixing MasterComp with multiply_32x32_rshift32

m-m-adams committed:
bugfix for MPE CC and global commands (#201)

stellar-aria committed:
Add generic container specializations (#202)

* Add generic container specializations

* Fix comment for queue

m-m-adams committed:
Create a q31 fixed point alias for int32_t in DSP code (#207)

* Create q31 typedef and move Q31 int instructions to new file

* Change order of asm operands and format to match

* replace magic number with constant and use fixed point typedef

* resolve missed conflict

stellar-aria committed:
Integrate SCons and Tasks into unified DBT (#206)

Update GitHub Actions

stellar-aria committed:
Fix error message (#208)

This fixes a bug where the error message for an invalid command would not display properly and instead fail

m-m-adams committed:
Fix for missing menu items on 7seg (#213)

* fix for missing menu items on 7seg

dbt format

* address review from stellar-aria

alter-alter committed:
Fix: Stop audio preview from the sample browser when entering Manual Slicer mode. (#210)

m-m-adams committed:
Svf fixes (#212)

* Enable 2x oversampling to fix stability issues at high cutoff

* set SVF parameters to more closely match ladder resonance and cutoff

* format fixes

bfredl committed:
Debug printing: refactor and allow USB output (#215)

* debug print: only use uartPrintLn in lowlevel code, otherwise use the c++ wrapper

* debug print: enable high-level debug output over USB as an option

request:
  F0 7D 03 00 [mode] F7
    mode = 00: off
    mode = 01: on (replies on the same midi device)
    otherwise reserved

reply:
  F0 7D 03 40 [kind] {ascii bytes 0-7f} F7
  kind = currently always 00, could be used for categories later

ascii bytes will use '\n' (0x0a) where a newline is intended.

* debug print: rename Uart::print to Debug::print

stellar-aria committed:
Propogate return code, add default build argument (#218)

alter-alter committed:
Update session_view.cpp (#220)

weavermedia committed:
Catch Notes setting to allow not catching existing notes. (#221)

* Add on/off setting for Catch Notes to allow not catching late notes.

* Formatting

---------

Co-authored-by: weavermedia <dan@Dweaver.media>

stellar-aria committed:
Refactor definitions.h for modern C++ (#219)

* Split up C and C++ definitions

* enum class MidiLearn

* enum class InstrumentType

* enum class ThingType

* enum class EnvelopeStage

* enum class PatchSource

* enum class LFOType

* enum class OscType

* Class member conversion

* enum class AudioInputChannel

* enum class SynthMode

* enum class ModFXType

* enum class SampleRepeatMode

* enum class FilterType

* enum class ArpMode

* enum class LPFMode

* enum class ModFXParam

* enum class GlobalMIDICommand

* enum class DrumType

* enum class AudioFileType

* enum class SequenceDirection

* enum class KeyboardLayout

* enum class AudioRecordingFolder

* enum class ActionResult

* enum class ScaleType

* enum class ArmState

* enum class MIDITakeoverMode

* enum class InputMonitoringMode

* various enum classes

* enum class Availability

* constexpr conversion and Cartesian buttons

* enum class IndependentNoteRowLengthIncrease

* enum class ClusterType, InterpolationMode

* enum class MarkerType

* Formatting

* Convert constants in definitions to proper style

* Formatting

m-m-adams committed:
Fix for i008 in latest community (#228)

* break macro into function to allow breakpoints

* remove subtraction by one, change to enum meant that this should now be the total capacity

bfredl committed:
sound: unison stereo spread (#223)

For now the shortcut is to press SELECT inside the unison number menu,
but we could also take the free spot above it (right of decimation).

m-m-adams committed:
explicitly access fakeparams to fix optimization error (#230)

m-m-adams committed:
refactor filter configs into seperate classes (#233)

soymonitus committed:
Fix the Doxyfile so it looks up recursively for all sources to generate full documentation (#235)

soymonitus committed:
Batch delete all “unused” kit rows (the ones with no notes) (#234)

* Batch delete all “unused” kit rows (the ones with no notes)

This is a big deal for me when I load an entire folder of percussion samples, play around with them till I find maybe a few that I like and create a sequence. Then later I want to quickly delete all the other rows that I didn’t use. I am using the Shift + Delete shorcut which was unused now, while in Kit view

* Add community features menu option

* Show popup if no unused rows to delete

* Change combo from just Shift+Delete to Kit+Shift+Delete

* Changed no notes warning text

* Changed the command to the more intuitive “Note press + Affect Entire + Delete”

- Added also a  “Deleted unused rows” popup.
- Removed the now unneeded  “At least one row needs to have notes” popup because that condition can’t happen (if you hold a note in an empty row you will unavoidably add a new note, making that row not empty, so it is impossible that all rows are empty when you initiate this command

* Revert "Changed the command to the more intuitive “Note press + Affect Entire + Delete”"

This reverts commit 901fe6b503f47fe55f60fc224f7b6dcd40b69478.

* Add popup to make it clear what happened when pressing the combo

* Trivial

* Added info to CommunityFeatures documentation

m-m-adams committed:
Bugfix mpe kit handling of CCs, pitch bend, and aftertouch (#236)

* git blame ignore file to remove the reformat from blame results

* fix mpe zone check in learned midi and switch kit to use that check

bfredl committed:
build: don't do -flto on debug builds (#238)

assuming this was a mistake. `-flto` is mentioned specifically
among the release flags, but got duplicated amongst the shared
flags, so it got applied to debug builds.

Having `-flto` unconditionaly in LINKFLAGS is fine. If there is
no LTO data in the .o files, the linker will just skip it.

stellar-aria committed:
Enable dbt format to alphabetize includes (#237)

* Update clang-format and `dbt format`

* Run format

* Fix build

* Resolve review comments

* Fix minor bug with dbt format

stellar-aria committed:
Hotfix for when dbt format is called without args (#246)

sapphire-arches committed:
Github pages: init (#248)

* workflow: only run when reasonable

this should prevent spurious runs on commits that should only rebuild the pages deployment

* workflow: build-pages init

* pages: better error message on fetch failure

* workflows: trigger on self changes

* workflow fixup scratch

* pages: temp fetch from bobtwinkles

This should get the owner/repo from the environment somehow instead
probably

* pages: fix fetch-artifacts on GH actions

* pages: fix overflow scroll looking weird on chrome

* page: artifact retrieval logging

* page: render relative time client side

* pages: fetch firmware images as part of next build

* pages: move back to root folder

* workflows: trigger pages rebuild on build completion

* pages: fix path in workflow

* workflows: remove path filters from required checks

* pages: remove references to old fetch script

* pages: pull repo from environment in workflow

* pages: note that recent PR builds are not ready

* pages: put commit description on download page

also emphasize this is community firmware in the header

* page: fix commit alignment css

stellar-aria committed:
Refactor out CacheManager from MemoryRegion (#209)

* Factor out CacheManager

* Refactor freeSomeStealableMemory

* Formatting

PaulFreund committed:
Performance keyboard layout API and in-key only layout (#138)

* Added first mention of extended keyboard view

* Added first mention of extended keyboard view

* Started structuring for layouts

* Started working on layout API

* Continue with cleanup

* Further commenting and refactoring, extension of layoutAPI

* Further work on the API and first integration of isomorphic layout as the first static call to test migration work

* More refactoring and comments

* [WIP] Midway padAction refactoring

* [WIP] Mostly figured out separation between Layout and KeyboardScreen

* Only yDisplay and color (re)calculation left to analyze

* [WIP] Further refactoring

* [WIP] Continued rework, made compilable again

* Ran ./dbt format src

* Formatting fixes

* Figured out most of note enable/disable signalling

* I can sound notes again \o/

* Fixed setting root note

* Little cleanup before starting rendering

* Keyboard is glowing and scrollable again (with missing bounds check right now)

* Mini cleanup

* Fully ported isomorphic keyboard except

* Mocked velocity_drums for now and added scrolling through layouts

* Re-Added highlighting selected pads in Isomorphic

* Cycle only through supported layouts and search for the next one if necessary

* Make sure layout is valid on opening

* Mini cleanup and notes

* State encapsulation per layout and saving current layout

* Finished velocity drums

* Added copied In-Key layout and fixed drum layout for Instruments

* Refactoring

* Added isKit API function for later

* Ensure recalculation of scroll bounds on clip changes

* Finished description

* [WIP] InKey layout

* Reverted calculation change in isomorphic

* Some octaves already working but not quite

* Happy with the colors, only upper scrolling limit left

* Fixed underflow potential

* Finished In-Key keybaord :)

* Forgot to implement storage - whoop, now really ready!

* Small capitalization fix to update GitHub PR issue

weavermedia committed:
Add Catch Notes feature to CommunityFeatures.md. Update feature list. General tidy up. (#253)

Co-authored-by: weavermedia <dan@Dweaver.media>

m-m-adams committed:
Improve 7seg text scroll and edit behaviour (#254)

* add default false option to stop the warning

* start scroll from difference in previous line

* refactor to remove gotos

* show end of line when saving with automatically incremented number

* provide additional context

* clang format

stellar-aria committed:
Universal DBT building via CMake (#252)

* CMake working

* Update for current source tree

* Flag parity with e2

* Source changes (include fixup/copy define)

* Update CMake config

* Update DBT tasks

* Add option to surpress python upgrade to dbt.cmd

* e2 build config via Makefile

* Update .gitignore

* CMake build Tasks

* Set generator in VSCode CMake settings

* CMakeLists.txt population

* Update target name

* Add `dbt --help`

* Add `dbt nuke` task for build dir removal

* Update GitHub build actions

* Add buildgen prelude to CMakeLists.txt

* Update README

* Add version and metadata tag option

* Fix build, update README

* Update VSCode files

* VSCode tasks problem matcher

* e2 project filters

* Fix string bug with dbt configure

* Hopefully fix build

* Globbed source lists

* Move some things to more logical locations

* Move firmware targets back to toplevel

m-m-adams committed:
Revert "Universal DBT building via CMake (#252)" (#258)

This reverts commit 5f80c5c4f6f2143d5b063b11f871dcbad26e8cba.

stellar-aria committed:
e2 via CMake (#265)

* CMake working

* Update for current source tree

* Flag parity with e2

* Source changes (include fixup/copy define)

* Update CMake config

* Update DBT tasks

* Add option to surpress python upgrade to dbt.cmd

* e2 build config via Makefile

* Update .gitignore

* CMake build Tasks

* Set generator in VSCode CMake settings

* CMakeLists.txt population

* Update target name

* Add `dbt --help`

* Add `dbt nuke` task for build dir removal

* Update GitHub build actions

* Add buildgen prelude to CMakeLists.txt

* Update README

* Add version and metadata tag option

* Fix build, update README

* Update VSCode files

* VSCode tasks problem matcher

* e2 project filters

* Fix string bug with dbt configure

* Hopefully fix build

* Globbed source lists

* Move some things to more logical locations

* Move firmware targets back to toplevel

* Fix issue with linker script (hopefully the last time)

* Embed firmware version in firmware file

m-m-adams committed:
UX guidelines (#222)

* first draft of UX guide

* address review comments from weavermedia and Jamie

* formatting

Co-authored-by: Jeff Way <jeff.way@me.com>

* Review from trappar

* rename to snake case

* incorporate feedback from @weavermedia and @PaulFreund

---------

Co-authored-by: Jeff Way <jeff.way@me.com>

sapphire-arches committed:
workflows: run less often (#272)

This should prevent the double-run of workflows when the merge queue
kicks code in to `community`. Requires some modification of the "gather
all runs for community commits" logic in the pages build.

It should also fix the bug where it's not possible to manually run the
pages deploy.

sapphire-arches committed:
tasks: add check mode to format task (#273)

passing --check (or -c) to the format check now makes the task not
modify the files, but rather use clang-format's check mode to report an
error and exit with a nonzero code.

this will be handy for those looking to set up pre-commit hooks that
error out on format check failures

m-m-adams committed:
Fix bipolar display of pan (#270)

* fix bipolar display of pan by marking virtual const override functions

* clang format

m-m-adams committed:
Save empty song UI (#267)

* allow folders on first folder read if no song name set

* update debug launch configs for cmake

stellar-aria committed:
Turn the GeneralMemoryAllocator into a singleton (#268)

* Turn GMA into a singleton

* Actually remove the global.

stellar-aria committed:
Standard types, int32_t, and min/max (#278)

* Replace custom typedefs with standard defs

* Change r_typedef to cstdint in deluge folder

* Replace getMin and getMax with strongly-typed vers

* Convert int to strict int32_t

* Format

sapphire-arches committed:
Document FINE TEMPO (#276)

Followup for #178

sapphire-arches committed:
pages: also collect artifacts for PRs (#279)

stellar-aria committed:
Menu Item cleanup (#190)

* Move selection options to `Sized`, Toggle class

* change basicTitle to title; place in constructors

* remove soundEditor.currentValue

Finish menu item tidying

* Fix Toggle scroll

* Fix nullptr and signed/unsigned bugs

* Turn Enumeration back to int from size_t

* Boundary check for selection list drawing

* Fix drawItemsForOLED

* Fix last found bugs before continuing refactor

* Move to static_vector

* Compiling with static_vector and templates

* Move Title to deluge::string

* Fix some bugs

* Add fmt as submodule

* Stringify and runtime-formatted titles

* building for both

* tweak static_vector

* Building for both

* Building on both

* Format includes

* Remove fmt submodule

* Working build after merge

* Typed Enumeration and Selection

* Fix lookup for getMax/MinValue in Param

* Run format

* Finish conversion to string

* A little bit of cleanup

* Swap allocator (and new/delete) to delugeAlloc

* Format

* Format

stellar-aria committed:
Hotfix for building on Windows (#280)

soymonitus committed:
Alternative golden knob delay params (ability to change encoder click behavior) (#282)

* Alternative parameters can be edited by pressing Delay golden knobs

Instead of changing PingPong On/Off and Type Analog/Digital, they will change SyncType and SyncLevel

* Cycle from Whole to 256th (omit sync_level_none)

* Update docs with correct PR url

trappar committed:
Add documentation for quantize feature (#283)

* Add documentation for quantize feature

* Update docs based on PR feedback

sapphire-arches committed:
task-build: support relwithdebuginfo (#284)

sapphire-arches committed:
chore: make sample rate a constant (#285)

This lets us document exactly why the sample rate is fixed to 44100, and
makes a number of computations clearer. There's a few more magic numbers
that are derived from 44100 in an unclear manner so I didn't touch
those. There's also a 44100 in r_usb_hhub_sys.c but it's hard to plumb a
c++ constant in to that C code so I opted to not worry about it for now.
Eventualyl we'd probably like to pull all of the changes out of the
Renesas code anyway, or at least have a cleaner interface.

sapphire-arches committed:
alloc: actually returned allocated memory (#286)

sapphire-arches committed:
keyboard-view: use const names (#287)

* keyboard-view: use const names

this avoids a -Wwrite-strings warning

also adds -Wwrite-strings as an error in debug builds, so we don't
accidentally commit this particular sin again

* ui: fix Wwrite-string warning in delete_file ctx menu

soymonitus committed:
Make Alternative Golden Knob Delay params community feature Off by default as it changes default behaviour (#290)

This is a followup to PR https://github.com/SynthstromAudible/DelugeFirmware/pull/282

PaulFreund committed:
Added reset commands fixes load sometimes not working (#289)

sapphire-arches committed:
ui: use string_view for title (#288)

this fixes a warning where we were returning a reference to a temporary
which is generally no good.

The total firmware cost is ~10k additional bytes but that's probably
because the optimizer was stripping important code away

stellar-aria committed:
Fix debugging and add linker error parsing for e2 (#291)

bfredl committed:
refactor: move debug sysex code to own file

bfredl committed:
feat: load firmware over SysEX

stellar-aria committed:
Switch to format_string, decrease fmtlib compiled size (#294)

stellar-aria committed:
Migrate deluge::string to std::string (#302)

stellar-aria committed:
Merge Typed and untyped Enumeration and Selection MenuItems (#303)

* Combine Typed and Untyped Enum/Selection

* camel case getter/setter

m-m-adams committed:
Define relwithdebinfo target (#306)

* update relwithdebinfo target

enable lto in relwithdebinfo to stop crash in voice stealing

format fix

* correct test to relwithdebinfo

alter-alter committed:
Fixed Song Change MasterComp Setting Issue (#308)

Fixed the issue where swapping to a different song during playback could lead to unintended changes in the MasterComp settings before the start of the next song's playback.

bfredl committed:
fix: allow load+synth and save+synth again in keyboard view (#307)

trappar committed:
Only run pages deploy on community branch (#313)

alter-alter committed:
Resolved Audition Pad Illumination Issue on Drum Keyboard Screen (#317)

In the Drum Keyboard screen, we have fixed the issue where audition pads would remain illuminated after pressing pads 9 to 16 and then returning to the clip view.

sapphire-arches committed:
ub: Remove illegal FAM from ParamSet (#299)

Flexible Array Members are not legal in C++ and their use here has
already caused at least one "miscompilation" (see #230). This change
replaces it with regular arrays in the children, with a pointer passed
up to the parent. This costs about 1k extra bytes, but that seems likely
to be code that was being compiled out when it shouldn't be.

Baseline (5edc967): 1540416 build/Release/delugeOLED.bin
With this change  : 1541344 build/Release/delugeOLED.bin

stellar-aria committed:
Add nightly builds to github actions

Fixups from feedback

stellar-aria committed:
Merge branch 'community' into feature/nightly-build

jamiefaye committed:
Merge pull request #319 from stellar-aria/feature/nightly-build

Add nightly builds to Github actions
